B2B inventory 16-4

pfertyk opened the pull request:
Description of the issue/feature this PR addresses:
* cleaned up batch picking views
* made barcode unique only within a company
* improved purchase and purchase requisition views
* allowed changing operation type in "draft" for stock pickings
* appended "copy" to duplicated resources
* fixed long names display in product labels
* improved stock and stock picking views

Current behavior before PR:

Desired behavior after PR is merged:




---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/122511).

ticodoo reviewed (commented):
So far it looks fine :)

Some tips:
- for commits that undo something from another commit, you can make them fixup commits so it's more obvious + it will make the commit cleaning easier later on
- the commits you tagged with [REM] (that aren't fixups) should be [IMP] since they're more of improvements rather than the things [REM] are traditionally used for: 
`[REM] for removing resources: removing dead code, removing views, removing modules, …;`
- I forgot to mention that you should include a Task: {id# from the project task} in your commits. I can help you with this if you don't know what this means

pfertyk commented on addons/stock/report/report_lot_barcode.xml:
```diff
@@ -25,7 +25,9 @@
                                         <t t-elif="o.product_id.tracking == 'serial'" t-set="final_barcode" t-value="(final_barcode or '') + '21' + o.name"/>
                                     </t>
-                                    <div class="o_label_4x12" t-field="o.product_id.display_name" t-att-style="'width:22mm' if final_barcode else ''"/>
+                                    <!--Product's display_name has to be split in 2 lines, so we're using default_code and name separately -->
```
OK :) I don't have a habit of checking detailed commit messages (yet), so I preferred to put this in a comment, which I know will be read ;) I also wasn't super happy about duplicating the logic from our model (where we combine the reference and product name), so I felt the need of explaining this somewhere ;)

clesgow commented on addons/stock/report/report_lot_barcode.xml:
```diff
@@ -25,7 +25,8 @@
                                         <t t-elif="o.product_id.tracking == 'serial'" t-set="final_barcode" t-value="(final_barcode or '') + '21' + o.name"/>
                                     </t>
-                                    <div class="o_label_4x12" t-field="o.product_id.display_name" t-att-style="'width:22mm' if final_barcode else ''"/>
+                                    <div class="o_label_4x12" t-att-style="'width:22mm' if final_barcode else ''">[<span t-field="o.product_id.default_code"/>]</div>
```
```suggestion
                                    <div class="o_label_4x12" t-att-style="'width:22mm' if final_barcode else ''">[<t t-out="o.product_id.default_code"/>]</div>
```
Having a `<span>` here creates an unwanted padding.

Also, this raises issues with the display of a classic barcode:
![image](https://github.com/odoo/odoo/assets/18055894/d3da379e-acc2-44fd-845f-5130f076b3de)

pfertyk replied:
Changed `span` to `t` :+1:

clesgow commented on addons/product/models/product_product.py:
```diff
@@ -10,6 +10,16 @@
 from odoo.tools import float_compare
 
 
+def _get_barcodes_by_company(products_recordset):
+    barcodes_by_company = defaultdict(list)
+
+    for p in products_recordset:
+        if p.barcode:
+            barcodes_by_company[p.company_id.id].append(p.barcode)
+
+    return barcodes_by_company
```
No reasons for this to be outside of the class, move it inside and use `self`.
```suggestion
def _get_barcodes_by_company(self):
    barcodes_by_company = defaultdict(list)

    for product in self:
        if product.barcode:
            barcodes_by_company[product.company_id.id].append(product.barcode)

    return barcodes_by_company
```

pfertyk replied:
Moved the method into the class :+1:

clesgow reviewed (commented):
As a global remark:
For your commits, you don't need to explain *what* you did (since we can pretty much see it in the diff). However, you need to explain *why* this specific change is done in every commit. In the future, these commits message will inform you when you're about to change a line *why* it was written that way in the first place. :wink: 
Also, try to keep your commit titles under 70 characters to avoid an annoying line break on them, makes it all more readable this way.
This can be found in our [Git guidelines](https://www.odoo.com/documentation/master/contributing/development/git_guidelines.html#commit-message-full-description). :eyes:

pfertyk commented:
Hi @clesgow , thank you for your feedback :+1: I've already addressed some of the comments, with 2 issues remaining:
* about commit messages: should I update all of them in this PR to reflect the "why, not what" principle?
* about pickings in batch: I will clarify this with the PO and get back to you

clesgow commented on addons/stock_picking_batch/views/stock_picking_batch_views.xml:
```diff
@@ -157,7 +141,7 @@
                             <field name="allowed_picking_ids" invisible="1"/>
                             <field name="picking_ids" widget="many2many" mode="tree,kanban"
-                                context="{'form_view_ref': 'stock_picking_batch.view_picking_form_inherited', 'tree_view_ref': 'stock_picking_batch.view_picking_tree_batch'}"/>
+                                context="{'form_view_ref': 'stock_picking_batch.view_picking_form_inherited'}"/>
```
After check with the PO: Yes we do.

amoyaux commented on addons/product/models/product_product.py:
```diff
@@ -184,27 +184,53 @@ def init(self):
-        to ensure the uniqueness between products' barcodes and packagings' ones"""
-        all_barcode = [b for b in self.mapped('barcode') if b]
-        domain = [('barcode', 'in', all_barcode)]
+    def _get_barcodes_by_company(self):
```
check if it could be replace by groupby in odoo/tools/misc.py

amoyaux commented on addons/product/models/product_product.py:
```diff
@@ -184,27 +184,53 @@ def init(self):
-            duplicates_as_str = "\n".join(
-                _("- Barcode \"%s\" already assigned to product(s): %s", barcode, ", ".join(p.display_name for p in products))
-                for barcode, products in products_by_barcode.items() if len(products) > 1
+
```
optimize the only line that didn't change
matched_products = self.sudo()._search(domain)

pfertyk replied:
Here I used `read_group` combined with `search`. I understand that the purpose was to improve performance, and I think it works in this regards, as we don't perform the search unless we have a list with more than one item. I tried to use `browse` instead of `search`, but there were some problems (full error log is a bit too long to post here).

Please let me know if this solution is good enough or if you want to discuss it further @amoyaux .

pfertyk replied:
@amoyaux  I guess we can keep the current solution that uses `search` but in a much more limited way (only searching if there are duplicated barcodes)?

amoyaux commented on addons/stock/views/stock_picking_views.xml:
```diff
@@ -207,8 +207,7 @@
                             <field name="picking_type_id" options="{'no_open': True}"
-                                   attrs="{'invisible': [('hide_picking_type', '=', True)], 'readonly': [('id', '!=', False)]}"
-                                   domain="context.get('restricted_picking_type_code', []) and [('code', '=', context.get('restricted_picking_type_code'))]"/>
+                                   attrs="{'invisible': [('hide_picking_type', '=', True)], 'readonly': ['&amp;', ('state', '!=', 'draft'), ('id', '!=', False)]}"/>
```
Not sure about the feature. We will see with THD

pfertyk replied:
As discussed, I've reverted my changes in the search domain. However, I kept the condition allowing us to edit the operation type while in draft. I'll inform THD about this.

pfertyk replied:
THD confirmed that this is OK.

amoyaux commented on addons/stock/models/stock_location.py:
```diff
@@ -444,6 +444,12 @@ class StockRoute(models.Model):
         'Warehouses', copy=False, domain="[('id', 'in', warehouse_domain_ids)]")
 
+    def copy(self, default=None):
+        default = dict(default or {})
```
self could contains multiple records.

svs-odoo commented on addons/stock/models/stock_location.py:
```diff
@@ -444,6 +444,12 @@ class StockRoute(models.Model):
+    def copy(self, default=None):
+        default = dict(default or {})
+        if 'name' not in default:
+            default['name'] = _("%s (copy)") % self.name
```
```suggestion
            default['name'] = _("%s (copy)", self.name)
```

Or even like this if you fell fancy :monocle_face: 
```suggestion
            default['name'] = _("%(location_name)s (copy)", location_name=self.name)
```
This later version is useful when there is multiple parameters (to be able to change the position of the terms if it's appropriate regarding the lang), with only one parameter, it only gives more context for the translators, indicates to them what they are currently translating

pfertyk replied:
:+1:

pfertyk replied:
Changed in all 4 places ;)

svs-odoo reviewed (changes requested):
Hello !
This branch need to be rebased (I guess the conflicts result of the removing of the `attrs` in the views)

pfertyk commented on addons/stock/report/report_lot_barcode.xml:
```diff
@@ -25,7 +25,8 @@
                                         <t t-elif="o.product_id.tracking == 'serial'" t-set="final_barcode" t-value="(final_barcode or '') + '21' + o.name"/>
                                     </t>
-                                    <div class="o_label_4x12" t-field="o.product_id.display_name" t-att-style="'width:22mm' if final_barcode else ''"/>
+                                    <div class="o_label_4x12" t-att-style="'width:22mm' if final_barcode else ''">[<t t-out="o.product_id.default_code"/>]</div>
```
Good catch, thanks!

pfertyk commented on addons/stock_picking_batch/views/stock_picking_views.xml:
```diff
@@ -14,6 +14,16 @@
                         attrs="{'invisible': [('batch_id', '=', False)]}"/>
             </div>
+            <xpath expr="//page[@name='extra']//field[@name='company_id']" position="after">
+                <field name="batch_id"
```
Changed, thanks! And thank you again for explaining how views are processed ;)

amoyaux commented:
robodoo r+ rebase-ff

robodoo commented:
Merge method set to rebase and fast-forward.

robodoo commented:
@pfertyk @amoyaux linked pull request(s) odoo/upgrade#5138 not ready. Linked PRs are not staged until all of them are ready.

amoyaux commented on addons/stock_picking_batch/views/stock_picking_views.xml:
```diff
@@ -21,7 +21,7 @@
                             ('picking_type_id', '=', False),
                     ]"
-                    readonly="state in ['cancel', 'done']"
+                    readonly="1"
```
don't forget to remove the domain in that case

pfertyk replied:
Domain removed, thanks!

pfertyk commented on addons/product/models/product_product.py:
```diff
@@ -184,27 +184,53 @@ def init(self):
+
+        products_by_barcode = defaultdict(list)
+        for product in matched_products:
+            products_by_barcode[product.barcode].append(product)
```
User `read_group` instead, works like a charm :) Thanks!

amoyaux commented:
robodoo r+ rebase-ff

robodoo commented:
Merge method set to rebase and fast-forward.
